File: yunq/parser.py
from enum import Enum
import sys

class LexemeType(Enum):
    NONE = 0

    EOF = 1

    # Identifiers and Keywords
    NAME = 2
    
    # Symbols
    LEFT_BRACE = 3
    RIGHT_BRACE = 4
    LEFT_PAREN = 5
    RIGHT_PAREN = 6
    ARROW = 7
    SEMICOLON = 8
    DOT = 9


class Lexeme():
    def __init__(self, lextype: LexemeType, value = None):
        self.type = lextype
        self.value = value

    def __str__(self):
        if self.value:
            return "(%s, %s)" % (self.type, self.value)
        return "(%s)" % self.type

    def __repr__(self):
        return self.__str__()


def lexer(program: str):
    line = 1
    start = 0
    current = 0
    tokens: list[Lexeme] = [] 
    while current < len(program):
        # Scan next token.
        start = current
        curr = program[current]
        if curr == '\n':
            line += 1
        elif curr == '\t' or curr == ' ' or curr == '\r':
            pass
        elif curr == '{':
            tokens.append(Lexeme(LexemeType.LEFT_BRACE))
        elif curr == '}':
            tokens.append(Lexeme(LexemeType.RIGHT_BRACE))
        elif curr == '(':
            tokens.append(Lexeme(LexemeType.LEFT_PAREN))
        elif curr == ')':
            tokens.append(Lexeme(LexemeType.RIGHT_PAREN))
        elif curr == ';':
            tokens.append(Lexeme(LexemeType.SEMICOLON))
        elif curr == '.':
            tokens.append(Lexeme(LexemeType.DOT))
        elif curr == '-':
            current += 1
            if program[current] == '>': 
                tokens.append(Lexeme(LexemeType.ARROW))
            else:
                sys.exit("Expected > after - got '%s' on line %d" % (program[current], line))
        elif curr.isalpha():
            while program[current + 1].isalnum() or program[current + 1] == '_':
                current += 1
            tokens.append(Lexeme(LexemeType.NAME, program[start:current + 1]))
        elif curr == '/' and program[current + 1] == '/':
            while program[current] != '\n':
                current += 1
        else:
            sys.exit("Got unexpected token %s on line %s." % (curr, line))

        current += 1

    tokens.append(Lexeme(LexemeType.EOF))

    return tokens

class Package():
    def __init__(self, names: list[str]):
        self.names = names

    def cpp_namespace(self):
        return "::".join(self.names)

class Method():
    def __init__(self, name: str, request: str, response: str):
        self.name = name
        self.request = request
        self.response = response

class Interface():
    def __init__(self, name: str, methods: list[Method]):
        self.name = name
        self.methods = methods

class Type(Enum):
    NONE = 0
    U64 = 1
    I64 = 2
    STRING = 3
    BYTES = 4
    CAPABILITY = 5

type_str_dict = {
        "u64": Type.U64,
        "i64": Type.I64,
        "string": Type.STRING,
        "bytes": Type.BYTES,
        "capability": Type.CAPABILITY,
        }

type_to_cppstr = {
        Type.U64: "uint64_t",
        Type.I64: "int64_t",
        Type.STRING: "glcr::String",
        Type.CAPABILITY: "z_cap_t",
        Type.BYTES: "glcr::Vector<uint8_t>"
        }

class Field():
    def __init__(self, fieldtype: Type, name: str, repeated = False):
        self.type = fieldtype
        self.name = name
        self.repeated = repeated

    def cpp_type(self):
        return type_to_cppstr[self.type]

class Message():
    def __init__(self, name: str, fields: list[Field]):
        self.name = name
        self.fields = fields

Decl = Interface | Message

name_dict: dict[str, Decl] = {}

class Parser():
    def __init__(self, tokens: list[Lexeme]):
        self.tokens = tokens
        self.current = 0

    def peektype(self) -> LexemeType:
        return self.tokens[self.current].type

    def peekvalue(self) -> str:
        return self.tokens[self.current].value

    def consume(self) -> Lexeme:
        self.current += 1
        return self.tokens[self.current - 1]

    def consume_identifier(self) -> str:
        tok = self.consume()
        if tok.type != LexemeType.NAME:
            sys.exit("Expected identifier got %s" % tok.type)
        return tok.value

    def consume_check(self, lex_type: LexemeType):
        tok = self.consume()
        if tok.type != lex_type:
            sys.exit("Expected %s got %s" % (lex_type, tok.type))

    def consume_check_identifier(self, name: str):
        tok = self.consume()
        if tok.type != LexemeType.NAME:
            sys.exit("Expected '%s' got a %s" % (name, tok.type))
        if tok.value != name:
            sys.exit("Expected '%s' got '%s'" % (name, tok.value))

    def parse(self) -> list[Decl]:
        decls = []
        while self.peektype() != LexemeType.EOF:
            decls.append(self.decl())
        return decls

    def decl(self) -> Decl:
        token = self.consume()
        if token.type != LexemeType.NAME:
            sys.exit("Unexpected token: %s", token)
        if token.value == "package":
            # TODO: Enforce that package decl comes before all messages and interface.
            return self.package()
        if token.value == "message":
            return self.message()
        elif token.value == "interface":
            return self.interface()
        sys.exit("Unexpected identifier '%s', expected package, message, interface" % token.value)

    def package(self):
        names = [self.consume_identifier()]

        while self.peektype() == LexemeType.DOT:
            self.consume_check(LexemeType.DOT)
            names += [self.consume_identifier()]

        self.consume_check(LexemeType.SEMICOLON)

        return Package(names)

    def interface(self):
        # "interface" consumed by decl.
        name = self.consume_identifier()

        if name in name_dict.keys():
            sys.exit("Name '%s' already exists." % name)

        self.consume_check(LexemeType.LEFT_BRACE)

        methods: list[Method] = []
        method_names = set()
        while self.peektype() != LexemeType.RIGHT_BRACE:
            m = self.method()
            if m.name in method_names:
                sys.exit("Method %s declared twice on %s" % (m.name, name))
            method_names.add(m.name)
            methods.append(m)

        self.consume_check(LexemeType.RIGHT_BRACE)

        i = Interface(name, methods)
        name_dict[name] = i
        return i

    def method(self):
        self.consume_check_identifier("method")

        name = self.consume_identifier()

        self.consume_check(LexemeType.LEFT_PAREN)
        
        request = None
        # FIXME: Fix error handling here (and for response). We want to show
        # "expected rparen or identifier" if type is wrong rather than just 
        # "expected rparen".
        if self.peektype() == LexemeType.NAME:
            request = self.consume_identifier()

        self.consume_check(LexemeType.RIGHT_PAREN)
        self.consume_check(LexemeType.ARROW)
        self.consume_check(LexemeType.LEFT_PAREN)

        response = None
        if self.peektype() == LexemeType.NAME:
            response = self.consume_identifier()

        self.consume_check(LexemeType.RIGHT_PAREN)
        self.consume_check(LexemeType.SEMICOLON)

        return Method(name, request, response)

    def message(self):
        # "message" consumed by decl.

        name = self.consume_identifier()

        if name in name_dict:
            sys.exit("Name '%s' already exists." % name)

        self.consume_check(LexemeType.LEFT_BRACE)

        fields: list[Field] = []
        field_names = set()
        while self.peektype() != LexemeType.RIGHT_BRACE:
            f = self.field()
            if f.name in field_names:
                sys.exit("Field %s declared twice on %s" % (f.name, name))
            field_names.add(f.name)
            fields.append(f)

        self.consume_check(LexemeType.RIGHT_BRACE)

        m = Message(name, fields)
        name_dict[name] = m
        return m

    def field(self):

        repeated = False
        field_type_str = self.consume_identifier()
        if field_type_str == "repeated":
            repeated = True
            field_type_str = self.consume_identifier()
            
        if field_type_str not in type_str_dict.keys():
            sys.exit("Expected type got '%s'" % field_type_str)
        field_type = type_str_dict[field_type_str]

        name = self.consume_identifier()
        self.consume_check(LexemeType.SEMICOLON)
        return Field(field_type, name, repeated)

def type_check(decls: list[Decl]):
    if sum(1 for decl in decls if type(decl) is Package) > 1:
        sys.exit("Cannot have more than one package declaration")
    for decl in decls:
        if type(decl) is Interface:
            for method in decl.methods:
                if method.request is None and method.response is None:
                    sys.exit("Method '%s.%s' cannot have empty request and response" % (decl.name, method.name))
                if method.request is not None:
                  if method.request not in name_dict.keys():
                      sys.exit("Request type '%s' for '%s.%s' does not exist" % (method.request, decl.name, method.name))
                  if type(name_dict[method.request]) is not Message:
                      sys.exit("Request type '%s' for '%s.%s' should be a message" % (method.request, decl.name, method.name))
                if method.response is not None:
                  if method.response not in name_dict.keys():
                      sys.exit("Response type '%s' for '%s.%s' does not exist" % (method.response, decl.name, method.name))
                  if type(name_dict[method.response]) is not Message:
                      sys.exit("Response type '%s' for '%s.%s' should be a message" % (method.response, decl.name, method.name))

def print_ast(decls: list[Decl]):
    for decl in decls:
        if type(decl) is Package:
            print("%s (Package)" % decl.cpp_namespace())
        elif type(decl) is Interface:
            print("%s (Interface)" % decl.name)
            for method in decl.methods:
                print("\t%s (%s -> %s)" % (method.name, method.request, method.response))
        elif type(decl) is Message:
            print("%s (Message)" % decl.name)
            for field in decl.fields:
                print("\t%s %s" % (field.type.name, field.name))
        else:
            print("unknown type")

File: yunq/rust/src/parser.rs
use std::collections::HashMap;
use std::collections::HashSet;
use std::fmt::Debug;
use std::fmt::Display;

use crate::lexer::Token;
use crate::lexer::TokenType;

#[derive(Clone)]
pub enum Type {
    U64,
    I64,
    String,
    Bytes,
    Capability,
    Message(String),
}

impl Type {
    pub fn rust_type(&self) -> &str {
        match self {
            Type::U64 => "u64",
            Type::I64 => "i64",
            Type::String => "String",
            Type::Bytes => "Vec<u8>",
            Type::Capability => "z_cap_t",
            Type::Message(s) => s,
        }
    }
}

impl Display for Type {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Type::U64 => "u64",
                Type::I64 => "i64",
                Type::String => "string",
                Type::Bytes => "bytes",
                Type::Capability => "cap",
                Type::Message(s) => s,
            }
        )
    }
}

impl TryFrom<&String> for Type {
    type Error = String;

    fn try_from(value: &String) -> Result<Self, Self::Error> {
        match value.as_str() {
            "u64" => Ok(Type::U64),
            "i64" => Ok(Type::I64),
            "string" => Ok(Type::String),
            "bytes" => Ok(Type::Bytes),
            "capability" => Ok(Type::Capability),
            _ => Ok(Type::Message(value.clone())),
        }
    }
}

#[derive(Clone)]
pub struct Field {
    pub field_type: Type,
    pub name: String,
    pub number: u64,
    pub repeated: bool,
}

#[derive(Clone)]
pub struct Message {
    pub name: String,
    pub fields: Vec<Field>,
}

pub struct Method {
    pub name: String,
    pub number: u64,
    pub request: Option<String>,
    pub response: Option<String>,
}

impl Debug for Method {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let request_str = match &self.request {
            Some(s) => format!("({})", s),
            None => "()".to_string(),
        };
        let response_str = match &self.response {
            Some(s) => format!("({})", s),
            None => "()".to_string(),
        };

        write!(
            f,
            "{}: {}: {} -> {}",
            self.number, self.name, request_str, response_str
        )
    }
}

pub struct Interface {
    pub name: String,
    pub methods: Vec<Method>,
}

pub enum Decl {
    Message(Message),
    Interface(Interface),
    Package(Vec<String>),
}

impl Debug for Decl {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Decl::Message(m) => {
                writeln!(f, "Message {}", m.name)?;
                for field in &m.fields {
                    let typestr = if field.repeated {
                        format!("repeated {}", field.field_type)
                    } else {
                        field.field_type.to_string()
                    };
                    writeln!(f, "\t{}: {} ({})", field.number, field.name, typestr)?;
                }
            }
            Decl::Interface(i) => {
                writeln!(f, "Interface {}", i.name)?;
                for method in &i.methods {
                    writeln!(f, "\t{:?}", method)?;
                }
            }
            Decl::Package(p) => {
                writeln!(f, "Package {};", p.join("."))?;
            }
        }
        Ok(())
    }
}

pub struct Parser<'a> {
    tokens: &'a Vec<Token>,
    current_index: usize,
    ast: Vec<Decl>,
    type_map: HashMap<String, Message>,
}

impl<'a> Parser<'a> {
    pub fn new(tokens: &'a Vec<Token>) -> Self {
        Self {
            tokens,
            current_index: 0,
            ast: Vec::new(),
            type_map: HashMap::new(),
        }
    }

    fn next_type_is(&self, tok_type: TokenType) -> bool {
        self.current_index < self.tokens.len()
            && self.tokens[self.current_index].token_type == tok_type
    }

    fn consume_token<'b>(&'b mut self) -> &'b Token {
        if self.current_index >= self.tokens.len() {
            panic!("Consumed tokens past end of input.")
        }

        let t = &self.tokens[self.current_index];
        self.current_index += 1;
        t
    }

    fn consume_token_type<'b>(&'b mut self, t: TokenType) -> Result<&'b Token, String> {
        let token = self.consume_token();
        if token.token_type == t {
            Ok(token)
        } else {
            Err(format!("Expected {:?} but found {}", t, token))
        }
    }

    fn consume_identifier<'b>(&'b mut self) -> Result<&'b Token, String> {
        self.consume_token_type(TokenType::Name)
    }

    fn consume_keyword<'b>(&'b mut self, keyword: &str) -> Result<&'b Token, String> {
        let token = self.consume_token_type(TokenType::Name)?;
        if token.chars != keyword {
            Err(format!("Expected '{}' but got {}", keyword, token))
        } else {
            Ok(token)
        }
    }

    fn package(&mut self) -> Result<Decl, String> {
        // 'package' was consumed by decl.
        let mut package_names = vec![self.consume_identifier()?.chars.clone()];
        while self.next_type_is(TokenType::Dot) {
            self.consume_token_type(TokenType::Dot)?;
            package_names.push(self.consume_identifier()?.chars.clone());
        }
        self.consume_token_type(TokenType::Semicolon)?;
        Ok(Decl::Package(package_names))
    }

    fn field(&mut self, number: u64) -> Result<Field, String> {
        let mut type_identifier = self.consume_identifier()?;
        let mut repeated = false;
        if type_identifier.chars == "repeated" {
            repeated = true;
            type_identifier = self.consume_identifier()?;
        }

        let parsed_type = Type::try_from(&type_identifier.chars)
            .map_err(|err| format!("{}: {}", err, type_identifier))?;
        let name_identifier = self.consume_identifier()?.chars.clone();

        self.consume_token_type(TokenType::Semicolon)?;

        Ok(Field {
            field_type: parsed_type,
            name: name_identifier,
            number,
            repeated,
        })
    }

    fn message(&mut self) -> Result<Decl, String> {
        // 'message' was consumed by decl.
        let name = self.consume_identifier()?.chars.clone();

        self.consume_token_type(TokenType::LeftBrace)?;

        let mut fields = Vec::new();
        let mut next_field_number = 0;
        while !self.next_type_is(TokenType::RightBrace) {
            fields.push(self.field(next_field_number)?);
            next_field_number += 1;
        }

        self.consume_token_type(TokenType::RightBrace)?;

        Ok(Decl::Message(Message { name, fields }))
    }

    fn method(&mut self, number: u64) -> Result<Method, String> {
        self.consume_keyword("method")?;

        let name = self.consume_identifier()?.chars.clone();

        self.consume_token_type(TokenType::LeftParen)?;
        let request = if self.next_type_is(TokenType::Name) {
            Some(self.consume_identifier()?.chars.clone())
        } else {
            None
        };
        self.consume_token_type(TokenType::RightParen)?;

        self.consume_token_type(TokenType::Arrow)?;

        self.consume_token_type(TokenType::LeftParen)?;
        let response = if self.next_type_is(TokenType::Name) {
            Some(self.consume_identifier()?.chars.clone())
        } else {
            None
        };
        self.consume_token_type(TokenType::RightParen)?;

        self.consume_token_type(TokenType::Semicolon)?;

        Ok(Method {
            name,
            request,
            response,
            number,
        })
    }

    fn interface(&mut self) -> Result<Decl, String> {
        let name = self.consume_identifier()?.chars.clone();

        self.consume_token_type(TokenType::LeftBrace)?;

        let mut methods = Vec::new();
        let mut next_method_number = 0;
        while !self.next_type_is(TokenType::RightBrace) {
            methods.push(self.method(next_method_number)?);
            next_method_number += 1;
        }

        self.consume_token_type(TokenType::RightBrace)?;

        Ok(Decl::Interface(Interface { name, methods }))
    }

    fn decl(&mut self) -> Result<Decl, String> {
        let token = self.consume_identifier()?;
        match token.chars.as_str() {
            "package" => self.package(),
            "message" => self.message(),
            "interface" => self.interface(),
            _ => Err(format!(
                "Expected one of 'package', 'message', or 'interface' but got {}",
                token
            )),
        }
    }

    pub fn parse_ast(&mut self) -> Result<(), String> {
        while !self.next_type_is(TokenType::EndOfFile) {
            let decl = self.decl()?;
            self.ast.push(decl);
        }
        Ok(())
    }

    pub fn ast(&'a self) -> &'a Vec<Decl> {
        &self.ast
    }

    fn ensure_message_and_interface_names_unique(&mut self) -> Result<(), String> {
        let mut names = HashSet::new();
        for decl in &self.ast {
            match decl {
                Decl::Message(m) => {
                    if names.contains(&m.name) {
                        // TODO: Keep token information for a better error message here.
                        return Err(format!("Duplicate name {}", m.name));
                    };
                    names.insert(m.name.clone());
                    self.type_map.insert(m.name.clone(), m.clone());
                }
                Decl::Interface(i) => {
                    if names.contains(&i.name) {
                        // TODO: Keep token information for a better error message here.
                        return Err(format!("Duplicate name {}", i.name));
                    };
                    names.insert(i.name.clone());
                }
                _ => {}
            }
        }
        Ok(())
    }

    fn type_check_message(&self, message: &Message) -> Result<(), String> {
        let mut field_names = HashSet::new();

        for field in &message.fields {
            if field_names.contains(&field.name) {
                return Err(format!(
                    "Field name '{}' used twice in message '{}'",
                    field.name, message.name
                ));
            }
            field_names.insert(field.name.clone());

            if let Type::Message(name) = &field.field_type {
                if !self.type_map.contains_key(name) {
                    return Err(format!(
                        "Unknown type '{}' on field '{}' in message '{}'",
                        name, field.name, message.name
                    ));
                }
            }
        }
        Ok(())
    }

    fn type_check_interface(&self, interface: &Interface) -> Result<(), String> {
        let mut method_names = HashSet::new();

        for method in &interface.methods {
            if method_names.contains(&method.name) {
                return Err(format!(
                    "Method name '{}' used twice in interface '{}'",
                    method.name, interface.name
                ));
            }
            method_names.insert(method.name.clone());

            if let Some(name) = &method.request {
                if !self.type_map.contains_key(name) {
                    return Err(format!(
                        "Unknown request type '{}' on method '{}' in interface '{}'",
                        name, method.name, interface.name
                    ));
                }
            }
            if let Some(name) = &method.response {
                if !self.type_map.contains_key(name) {
                    return Err(format!(
                        "Unknown response type '{}' on method '{}' in interface '{}'",
                        name, method.name, interface.name
                    ));
                }
            }
        }
        Ok(())
    }

    pub fn type_check(&mut self) -> Result<(), String> {
        self.ensure_message_and_interface_names_unique()?;
        for decl in &self.ast {
            match decl {
                Decl::Message(m) => self.type_check_message(m)?,
                Decl::Interface(i) => self.type_check_interface(i)?,
                _ => {}
            }
        }
        Ok(())
    }
}
